src/DefaultBuilder/src/StaticWebAssetsFileProvider.cs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.InteropServices;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore
+{
+    internal class StaticWebAssetsFileProvider : IFileProvider
+    {
+        private static readonly StringComparison FilePathComparison = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ?
+            StringComparison.OrdinalIgnoreCase :
+            StringComparison.Ordinal;
+
+        public StaticWebAssetsFileProvider(string pathPrefix, string contentRoot)
+        {
+            BasePath = pathPrefix.StartsWith("/") ? pathPrefix : "/" + pathPrefix;
+            InnerProvider = new PhysicalFileProvider(contentRoot);
+        }
+
+        public PhysicalFileProvider InnerProvider { get; }
+
+        public string BasePath { get; }
+
+        /// <inheritdoc />
+        public IDirectoryContents GetDirectoryContents(string subpath)
+        {
+            if (!StartsWithBasePath(subpath, out var physicalPath))
+            {
+                return NotFoundDirectoryContents.Singleton;
+            }
+            else
+            {
+                return InnerProvider.GetDirectoryContents(physicalPath);
+            }
+        }
+
+        /// <inheritdoc />
+        public IFileInfo GetFileInfo(string subpath)
+        {
+            if (!StartsWithBasePath(subpath, out var physicalPath))
+            {
+                return new NotFoundFileInfo(subpath);
+            }
+            else
+            {
+                return InnerProvider.GetFileInfo(physicalPath);
+            }
+        }
+
+        /// <inheritdoc />
+        public IChangeToken Watch(string filter)
+        {
+            return InnerProvider.Watch(filter);
+        }
+
+        private bool StartsWithBasePath(string subpath, out PathString rest)
+        {
+            return new PathString(subpath).StartsWithSegments(BasePath, FilePathComparison, out rest);
```

Comment:
Careful, the implicit PathString<->string converters change the encoding. BasePath needs to be a PathString too.

Reply:
Ah, could you spill the exact magic to use here? Maybe I can add a test that check encodings, but doesn't the url already arrive decoded?

Reply:
You don't need FilePathComparison, BasePath isn't considered as part of the file path, it's treated as a case insensitive prefix.
https://github.com/aspnet/AspNetCore/blob/08ed1afbbc1d1e11115eafcf4bd2c771c00ffd90/src/Middleware/StaticFiles/src/Helpers.cs#L40